wdl-lsp/src/proto.rs

```diff
@@ -84,3 +100,140 @@ pub fn diagnostic(
+    source: &str,
+) -> Option<DocumentDiagnosticReportResult> {
+    let result = results
+        .iter()
+        .find(|r| r.uri().as_ref() == &params.text_document.uri)?;
+
+    if let Some(previous) = params.previous_result_id {
+        if &previous == result.id().as_ref() {
+            log::debug!(
+                "diagnostics for document `{uri}` have not changed (client has latest)",
+                uri = params.text_document.uri,
+            );
+            return Some(DocumentDiagnosticReportResult::Report(
+                DocumentDiagnosticReport::Unchanged(RelatedUnchangedDocumentDiagnosticReport {
+                    related_documents: None,
+                    unchanged_document_diagnostic_report: UnchangedDocumentDiagnosticReport {
+                        result_id: previous,
+                    },
+                }),
+            ));
+        }
+
+        log::debug!(
+            "diagnostics for document `{uri}` have changed since last client request",
+            uri = params.text_document.uri
+        );
+    }
+
+    let items = result
+        .diagnostics()
+        .iter()
+        .map(|d| {
+            diagnostic(
+                result.uri(),
+                result
+                    .parse_result()
+                    .lines()
+                    .expect("should have line index"),
+                source,
+                d,
+            )
+        })
+        .collect::<Result<Vec<_>>>()
+        .ok()?;
+
+    Some(DocumentDiagnosticReportResult::Report(
+        DocumentDiagnosticReport::Full(RelatedFullDocumentDiagnosticReport {
+            related_documents: None,
+            full_document_diagnostic_report: FullDocumentDiagnosticReport {
+                result_id: Some(result.id().as_ref().clone()),
+                items,
+            },
+        }),
+    ))
+}
+
+/// Converts analysis results into an LSP workspace diagnostic report.
+pub fn workspace_diagnostic_report(
+    params: WorkspaceDiagnosticParams,
+    results: Vec<AnalysisResult>,
+    source: &str,
+) -> WorkspaceDiagnosticReportResult {
+    let ids = params
+        .previous_result_ids
+        .into_iter()
+        .map(|id| (id.uri, id.value))
+        .collect::<HashMap<_, _>>();
+
+    let mut items = Vec::new();
+    for result in results {
+        // Only store local file results
+        if result.uri().scheme() != "file" {
+            continue;
+        }
+
+        if let Some(previous) = ids.get(result.uri()) {
+            if previous == result.id().as_ref() {
+                log::debug!(
+                    "diagnostics for document `{uri}` have not changed (client has latest)",
+                    uri = result.uri(),
+                );
+
+                items.push(WorkspaceDocumentDiagnosticReport::Unchanged(
+                    WorkspaceUnchangedDocumentDiagnosticReport {
+                        uri: result.uri().as_ref().clone(),
+                        // TODO: FIX ME
```

Comment:
Should this be addressed before we merge?

Reply:
Whoops, I fixed it in one place but not the other. Fixing now.